table/tables/tables.go

```diff
@@ -1190,23 +1194,167 @@ func init() {
+	meta *model.SequenceInfo
+	// base < end when increment > 0.
+	// base > end when increment < 0.
+	end  int64
+	base int64
+	// round is used to count the cycle times.
+	round int64
+	mu    sync.Mutex
+}
+
+// GetSequenceBaseEndRound is used in test.
+func (s *sequenceCommon) GetSequenceBaseEndRound() (int64, int64, int64) {
+	return s.base, s.end, s.round
+}
+
 // GetSequenceNextVal implements util.SequenceTable GetSequenceNextVal interface.
-func (t *TableCommon) GetSequenceNextVal(dbName, seqName string) (int64, error) {
-	t.sequence++
-	// TODO: implements it with sequence allocation logic.
-	return t.sequence, nil
+// Caching the sequence value in table, we can easily be notified with the cache empty,
+// and write the binlogInfo in table level rather than in allocator.
+func (t *TableCommon) GetSequenceNextVal(dbName, seqName string) (nextVal int64, err error) {
+	seq := t.sequence
+	if seq == nil {
+		// TODO: refine the error.
+		return 0, errors.New("sequenceCommon is nil")
+	}
+	seq.mu.Lock()
+	defer seq.mu.Unlock()
+
+	var updateCache bool
+	err = func() error {
+		var err1 error
+		// Check if need to update the cache batch from storage.
+		// Because seq.base is not always the last allocated value (may be set by setval()).
+		// So we should try to seek the next value in cache (not just add increment to seq.base).
+		if seq.base == seq.end {
+			// There is no cache yet.
+			updateCache = true
+		} else {
+			// Seek the first valid value in cache.
+			offset := seq.meta.Start
+			if seq.meta.Cycle && seq.round > 0 {
+				if seq.meta.Increment > 0 {
+					offset = seq.meta.MinValue
+				} else {
+					offset = seq.meta.MaxValue
+				}
+			}
+			var ok bool
+			if seq.meta.Increment > 0 {
+				nextVal, ok = autoid.SeekToFirstSequenceValue(seq.base, seq.meta.Increment, offset, seq.base, seq.end)
+			} else {
+				nextVal, ok = autoid.SeekToFirstSequenceValue(seq.base, seq.meta.Increment, offset, seq.end, seq.base)
+			}
+			if !ok {
+				updateCache = true
+			}
+		}
+		if !updateCache {
+			return nil
+		}
+		// Update batch alloc from kv storage.
+		var sequenceAlloc autoid.Allocator
+		for _, alloc := range t.allocs {
+			if alloc.GetType() == autoid.SequenceType {
+				sequenceAlloc = alloc
+			}
+		}
+		if sequenceAlloc == nil {
+			// TODO: refine the error.
+			return errors.New("sequenceAlloc is nil")
+		}
+		seq.base, seq.end, seq.round, err1 = sequenceAlloc.AllocSeqCache(t.tableID)
+		return err1
+	}()
+	// Sequence alloc in kv store error.
+	if err != nil {
+		if err == autoid.ErrAutoincReadFailed {
+			return 0, table.ErrSequenceHasRunOut.GenWithStackByArgs(dbName, seqName)
+		}
+		return 0, err
+	}
+	// Sequence cache updated.
+	if updateCache {
+		// SeekToFirstAutoIDUnSigned seeks to next sequence value.
+		var ok bool
+		offset := seq.meta.Start
+		if seq.meta.Cycle && seq.round > 0 {
+			if seq.meta.Increment > 0 {
+				offset = seq.meta.MinValue
+			} else {
+				offset = seq.meta.MaxValue
+			}
+		}
```

Comment:
This lines is same with [L1239](https://github.com/AilinKid/tidb/blob/9547089e858d47f4c78e70c355ad5d2249a746c4/table/tables/tables.go#L1239-L1246), warp them to a function,
```golang
func (s *sequenceCommon) getOffset() int64 {
	offset := s.meta.Start
	if s.meta.Cycle && s.round > 0 {
		if s.meta.Increment > 0 {
			offset = s.meta.MinValue
		} else {
			offset = s.meta.MaxValue
		}
	}
	return offset
}
```

Reply:
nice catch.